### ISSUE TITLE
Add command to create a test case and jump between a tested entity and test case

### ISSUE THREAD

**Issue by JurajKubelka** [pharo-ide/Calypso#179]
Nautilus has a nice feature to create and jump to a test case. It can be done on a class and on a method.

I was playing with the same feature in Calypso and it looks like this for a selected class: 

![screen shot 2018-02-01 at 13 32 02](https://user-images.githubusercontent.com/465505/35691079-5283048e-0756-11e8-863a-50fb5e526f5f.png)
![screen shot 2018-02-01 at 13 31 24](https://user-images.githubusercontent.com/465505/35691065-486f98a4-0756-11e8-8825-d5ea2e0e6354.png)
![screen shot 2018-02-01 at 13 31 36](https://user-images.githubusercontent.com/465505/35691054-4303cc1e-0756-11e8-849e-9a770d98267b.png)

And similarly for a selected method: 

![screen shot 2018-02-01 at 13 32 31](https://user-images.githubusercontent.com/465505/35691164-82051d32-0756-11e8-903e-2a560c7408f9.png)
![screen shot 2018-02-01 at 13 32 19](https://user-images.githubusercontent.com/465505/35691165-83d1daf6-0756-11e8-83b3-22dcd0cfb9fb.png)

I switch between classes and methods using CMD+SHIFT+T. 
What do you think? Can we add this to the Calypso as a standard command? 
The code is not ready yet. I started with a collection of new commands. When I was almost done, I found out that very similar detection of test cases and tested classes and methods is already done in `Calypso-SystemPlugins-SUnit-Queries`. 

Actually my question is if it is worth to create a plugin, similar to `ClyTestedEnvironmentPlugin` to link test cases and tested classes/methods? 

Without such plugin, every time, those commands are asked for `canBeExecutedInContext:`, the link between tests and tested classes and methods has to be evaluated for each command. From this point of view, it is done just once while using a plugin. So it looks like good approach to create such plugin, right? 

Another question: Why `ClyTestedEnvironmentPlugin` is marked as slow?

So, this issue can stay opened if you agree to integrate those commands to the Calypso repository. If so, I will be happy to do it and prepare a pull request. If not, I will appreciate your comments about the best implementation :-)

The code is available here: 

```Smalltalk
Metacello new
  baseline: 'CalypsoExtensions';
  repository: 'github://JurajKubelka/CalypsoExtensions/src';
  load.
```

Note: the repository also includes some commands to create `initialize` and `accept:` methods. If you like them too, we can create another issue to integrate them.

**Comment by dionisiydk** [comment]
I will look in details later.

> Another question: Why ClyTestedEnvironmentPlugin is marked as slow?

Because of the way how related tests are trying to be found. It leads to full lookup of methods in many cases. It is fine in our machines. But for example on Raspberry dramatically breaks performance. Same happens with analysing override/overridden method status.

So I added this hook to be able disable all plugins which are known as slow.

**Comment by dionisiydk** [comment]
About shortcuts - I am not against. 
But I would like to use cmd+t to run tests in any context. If you shortcut do not conflict with it (I think it do not) I am fine.

**Comment by JurajKubelka** [comment]
The shortcut does no conflict. I choose CMD+SHIFT+T as it is close to the existing CMD+T. And thus easy to remember. Unfortunately, it appears as CMD+T in the the context menu. I have not investigate yet if it is a Calypso bug or Pharo bug.

**Comment by JurajKubelka** [comment]
According to the slowness: Does it mean that if I have a command that crates `initialize` method, it should not check in the `canBeExecutedInContext:` method if the `initialize` method already exists in a selected class? 

I do same for `accept:` method, where I also check if it is an abstract method.

**Comment by dionisiydk** [comment]
You example should work fine only on local environment. In case of remote browser each check for #initialize existence will produce remote call.

So my approach is to add special environment plugin which will mark classes that they have #initialize method. In remote scenario it is evaluated on server side when visible part of items are transferred to the client. 
For examples look how classes are marked with ClyAbstractItemTag. In your case it can be something like ClyClassInitializedTag. Or more complex property like ClyDefinedSelectorProperty with required selector as parameter.
Then Command can analyse it using: 
```Smalltalk
aContext lastSelectedItem isMarkedWith: ClyClassInitializedTag
```
or
```Smalltalk
property := aContext lastSelectedItem getProperty: ClyDefinedSelectorProperty ifAbsent: [^false].
^property selector == #initialize
```

**Comment by macta** [comment]
How can we unblock this one? Would like something similar and it sounds like it was very close, but 4 months later its still sitting in here...

**Comment by dionisiydk** [comment]
Hi Tim.
Problem that implementation in Nautilus was awfull. I did not want to not
copy this crap to Calypso.

But now if somebody will prepare dirty pull request I am fine. Better would
be of course to implement it in good fashion

пт, 22 июня 2018 г., 17:02 Tim Mackinnon <notifications@github.com>:

**Comment by macta** [comment]
Hi Denis - I'm trying to get a fork working so I can hope to contribute a bit - completely understand the desire not to tie yourself to Nautilus. Lets see what we can do.

**Comment by JurajKubelka** [comment]
Hi @dionisiydk: I have cleaned some GToolkit extensions according to your comments. There is one think that is not clear to me. Let's say that my plugin sets `ClyClassInitializedTag` (~ a class has `initialize` method). What is the right usage to listen to system changes and add or remove `ClyClassInitializedTag` according to class changes?

**Comment by dionisiydk** [comment]
Hi @JurajKubelka . 
In that case your plugin knows that classes need to be updated when method #initialize is added or removed. So the plugin should subscribe on MethodAdded and MethodRemoved events and detect relevant changes:
```Smalltalk 
YourEnvironmentPlugin>>attachToSystem
	super attachToSystem.
	
	environment system when: MethodAdded, MethodRemoved send: #processMethodChange: to: self.
```
Then you should introduce new system change like ClassInitializerChanged. When plugin will detect that #initialize was added or removed it will notify environment about this reified change:
```
YourEnvironmentPlugin>>processMethodChange: anEvent
anEvent methodAffected selector = #initialize ifTrue: [
        environment systemChanged: (ClassInitializerChanged inClass: anEvent classAffected overriddenMethod: method)]
```
ClassInitializerChanged can be subclass of ClassAnnouncement with required methods (probably just #classAffected). Class changes are already processed by calypso environment. So new type of class change should just work. It will force update of any class queries (which return classes) and therefore will force update of any class views
